plugins/parcelport/verbs/header.hpp

```diff
@@ -0,0 +1,185 @@
+//  Copyright (c) 2015-2016 John Biddiscombe
+//  Copyright (c) 2013-2015 Thomas Heller
+//  Copyright (c) 2013-2014 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_PARCELSET_POLICIES_VERBS_HEADER_HPP
+#define HPX_PARCELSET_POLICIES_VERBS_HEADER_HPP
+
+#include <hpx/runtime/parcelset/parcel_buffer.hpp>
+#include <hpx/util/assert.hpp>
+//
+#include <boost/array.hpp>
+//
+#include <cstdint>
+#include <cstddef>
+#include <cstring>
+#include <utility>
+#include <vector>
+
+// A generic header structure that can be used by parcelports
+// currently, the verbs parcelports make use of it
+namespace hpx {
+namespace parcelset {
+namespace policies {
+namespace verbs
+{
+    namespace detail {
+        struct rdma_region {
+            std::size_t  size;
+            uint32_t     key;
+            const void * addr;
+        };
+
+        struct piggy_back {
+            uint32_t  offset;
+            uint32_t  bytes;
+        };
+
+        union rdma_info {
+            piggy_back  piggyback;
+            rdma_region region;
+        };
+
+        typedef std::pair<uint32_t, uint32_t> num_chunks_type;
+
+        struct header_block {
+            uint32_t        tag;
+            uint32_t        size;
+            num_chunks_type num_chunks;
+            uint32_t        flags;
+            rdma_info       chunk_info;
+            rdma_info       message_info;
+        };
+    }
+
+    template <int SIZE>
+    struct header
+    {
+        static constexpr unsigned int header_block_size = sizeof(detail::header_block);
+        static constexpr unsigned int data_size_        = SIZE - header_block_size;
+        static const     unsigned int chunk_flag        = 0x01;
+        static const     unsigned int message_flag      = 0x02;
+        //
+    private:
+        detail::header_block           message_header;
+        boost::array<char, data_size_> data_;
```

Comment:
What's the difference between boost::array and std::array? Would it be possible to use the latter?